### PR TITLE
feat(server): Add JWT validation

### PR DESCRIPTION
**Pull request by aarushik93**
### Background

We are enabling authentication support for cloud versions. This PR adds a JWT validation PR, which decodes the related token and determines if it's valid & not expired. 

### Changes 🏗️
Added a new libs folder for shared functionality. 
Added auth middleware 
Added an environment variable check to enable or disable auth

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [ ] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

___

### **PR Type**
Enhancement, Documentation

___

### **Description**
- Added configuration settings for JWT authentication.
- Implemented JWT token parsing and validation utility.
- Added JWT authentication middleware for FastAPI.
- Integrated JWT authentication middleware into FastAPI server.
- Added README for shared libraries project.
- Defined project metadata and dependencies for autogpt-libs.
- Added example environment variable for enabling authentication.
- Added autogpt-libs dependency to server project.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>config.py</strong><dd><code>Add configuration settings for JWT authentication</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_libs/autogpt_libs/auth/config.py

<li>Added <code>Settings</code> class to manage JWT configuration.<br> <li> Loaded environment variables for JWT settings.<br> <li> Introduced <code>is_configured</code> property to check if JWT is properly <br>configured.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7642/files#diff-4f311f26904bfb1bfacffebd1a5b6688112789e77c04845585853ed1c8cc81f8">+16/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>jwt_utils.py</strong><dd><code>Implement JWT token parsing and validation utility</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_libs/autogpt_libs/auth/jwt_utils.py

<li>Added <code>parse_jwt_token</code> function to decode and validate JWT tokens.<br> <li> Handled token expiration and invalid token errors.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7642/files#diff-3602e96c161be5294c3321be632438da7a47a4da2d190044b43c39d99811d76d">+20/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>middleware.py</strong><dd><code>Add JWT authentication middleware for FastAPI</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_libs/autogpt_libs/auth/middleware.py

<li>Added <code>auth_middleware</code> function to handle JWT authentication.<br> <li> Integrated JWT token parsing and validation in middleware.<br> <li> Configured middleware to be conditional based on environment variable.<br> <br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7642/files#diff-f2419b83f733711998f92db2c33a2757ee0eccc6b63430f6730478e4a0415717">+26/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>server.py</strong><dd><code>Integrate JWT authentication middleware into FastAPI server</code></dd></summary>
<hr>

rnd/autogpt_server/autogpt_server/server/server.py

<li>Integrated <code>auth_middleware</code> into FastAPI server.<br> <li> Added dependency injection for authentication middleware.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7642/files#diff-94fec4bd2ec2c2e02a2e106e1ebcdce2d819c8c7fc0a30561d31fabb0e82d1ee">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Add README for shared libraries project</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_libs/README.md

<li>Added README for <code>autogpt_libs</code> project.<br> <li> Described purpose of shared libraries.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7642/files#diff-8a3863816ea3ccc00691a4990ef4e221650cad404e6fb93909a85b079caa9c72">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Define project metadata and dependencies for autogpt-libs</code></dd></summary>
<hr>

rnd/autogpt_libs/pyproject.toml

<li>Added project metadata for <code>autogpt-libs</code>.<br> <li> Defined dependencies for JWT and dotenv libraries.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7642/files#diff-5bb0a8bfa49ffda95ad4ef400a7754acfa8bae133726392a1dc2b29db93ad513">+17/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>.env.example</strong><dd><code>Add example environment variable for enabling authentication</code></dd></summary>
<hr>

rnd/autogpt_server/.env.example

- Added `ENABLE_AUTH` environment variable example.

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7642/files#diff-9dd1e3bfac62caa79a5a78c43140da9ea45d0d3f4d20f14b1fe894d6c40ce0af">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Add autogpt-libs dependency to server project</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_server/pyproject.toml

- Added `autogpt-libs` as a dependency.

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7642/files#diff-adb694b1ef2fb9dabbdd0efe486d08bc9336bcd7b824f13f4125228d70cd4d7a">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: rnd/autogpt_libs/autogpt_libs/auth/config.py
Comment (kcze) [inline rnd/autogpt_libs/autogpt_libs/auth/config.py:1]: No newline at the end

File: rnd/autogpt_libs/autogpt_libs/auth/middleware.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import logging
+
+from fastapi import Request, HTTPException, Depends
+from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
+from .jwt_utils import parse_jwt_token
+from .config import settings
+
+security = HTTPBearer()
+async def auth_middleware(request: Request):
```
Comment (kcze) [inline rnd/autogpt_libs/autogpt_libs/auth/middleware.py:9]: No newlines :)

File: rnd/autogpt_server/pyproject.toml [hunk 2]
```diff
@@ -40,6 +40,7 @@ feedparser = "^6.0.11"
 python-dotenv = "^1.0.1"
 expiringdict = "^1.2.2"
 
+autogpt-libs = {path = "../autogpt_libs"}
```
Comment (kcze) [inline rnd/autogpt_server/pyproject.toml]:
```suggestion
autogpt-libs = { path = "../autogpt_libs", develop = true }
```

**Review by kcze — APPROVED** [review]
LGTM

**Review by kcze — APPROVED** [review]

File: rnd/autogpt_libs/autogpt_libs/auth/jwt_utils.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import jwt
+from typing import Dict, Any
+from .config import settings
+
+
+def parse_jwt_token(token: str) -> Dict[str, Any]:
```
Comment (Swiftyos) [inline rnd/autogpt_libs/autogpt_libs/auth/jwt_utils.py:6]: Is it possible to be more specific about the type the token is?

File: rnd/autogpt_server/autogpt_server/server/server.py [hunk 4]
```diff
@@ -80,6 +82,8 @@ def run_service(self):
 
         # Define the API routes
         router = APIRouter(prefix="/api")
+        router.dependencies.append(Depends(auth_middleware))
```
Comment (Swiftyos) [inline rnd/autogpt_server/autogpt_server/server/server.py:85]: Do we have an example of how we would use this for access control?

Comment (aarushik93) [reply rnd/autogpt_server/autogpt_server/server/server.py:85]: Right now it's only authentication, ie authenticated users can access the API. RBAC and/or Scope based will come in follow up PRs, as I build out different user levels

File: rnd/autogpt_libs/autogpt_libs/auth/jwt_utils.py (hunk 3 above)
Comment (aarushik93) [reply rnd/autogpt_libs/autogpt_libs/auth/jwt_utils.py:6]: can you explain what you mean, please

**Pull request merged**